include/exadg/postprocessor/time_control.cpp

```diff
@@ -0,0 +1,126 @@
+#include <deal.II/base/exceptions.h>
+
+// ExaDG
+#include <exadg/postprocessor/time_control.h>
+
+
+TimeControl::TimeControl()
+  : EPSILON(1.0e-10),
+    reset_counter(true),
+    calculate(false),
+    start_time(std::numeric_limits<double>::max()),
+    calculate_every_time_interval(-1.0),
+    calculate_every_time_steps(std::numeric_limits<unsigned int>::max()),
+    counter(0),
+    last_timestep_number_to_evaluate(dealii::numbers::invalid_unsigned_int)
+{
+}
+
+void
+TimeControl::reinit(bool const         calculate_,
+                    double const       start_time_,
+                    double const       calculate_every_time_interval_,
+                    unsigned int const calculate_every_time_steps_)
+{
+  calculate = calculate_;
+
+  if(calculate_)
+  {
+    if(calculate_every_time_steps_ == std::numeric_limits<unsigned int>::max() ||
+       calculate_every_time_steps_ == 0)
+      AssertThrow(calculate_every_time_interval_ > 0.0 &&
+                    calculate_every_time_interval_ < std::numeric_limits<double>::max(),
+                  dealii::ExcMessage("Input parameters are in conflict."));
+    else
+      AssertThrow(calculate_every_time_interval_ < 0.0 ||
+                    calculate_every_time_interval_ >= std::numeric_limits<double>::max(),
+                  dealii::ExcMessage("Input parameters are in conflict."));
+  }
+
+  start_time                    = start_time_;
+  calculate_every_time_interval = calculate_every_time_interval_;
+  calculate_every_time_steps    = calculate_every_time_steps_;
+}
+
+// If reinit() is not called needs_evaluation() will always evaluate to false
+bool
+TimeControl::needs_evaluation(double const time, unsigned int const time_step_number) const
+{
+  if(calculate == false)
+    return false;
+
+  bool evaluate = false;
+
+  if(calculate_every_time_steps > 0)
+  {
+    if(time > start_time && (time_step_number - 1) % calculate_every_time_steps == 0)
+    {
+      evaluate = true;
+    }
+  }
+  else if(calculate_every_time_interval > 0.0)
+  {
+    // The current time might be larger than start_time. In that case, we first have to
+    // reset the counter in order to avoid that output is written every time step.
+    if(reset_counter)
+    {
+      counter += int((time - start_time + EPSILON) / calculate_every_time_interval);
+      reset_counter = false;
+    }
+
+    if((time > (start_time + counter * calculate_every_time_interval - EPSILON)))
+    {
+      evaluate = true;
+      ++counter;
+    }
+  }
+  else
+  {
+    AssertThrow(false,
+                dealii::ExcMessage(
+                  "Invalid parameters specified. Use either "
+                  "calculate_every_time_interval > 0.0 or calculate_every_time_steps > 0."));
+  }
+
+  if(evaluate)
+    last_timestep_number_to_evaluate = time_step_number;
+
+  return evaluate;
+}
+
+unsigned int
+TimeControl::get_counter() const
+{
+  return counter;
+}
+
+unsigned int
+TimeControl::get_last_timestep_number_to_evaluate() const
```

Comment:
I would like to get rid of this function.

Reply:
[d0bc416](https://github.com/exadg/exadg/pull/262/commits/d0bc416c46ebf80a9ec7966064f64303d03a538b)